src/jit/codegenarmarch.cpp

```diff
@@ -449,6 +449,10 @@ void CodeGen::genCodeForTreeNode(GenTree* treeNode)
         case GT_NOP:
             break;
 
+        case GT_SIDEEFFECT:
+            genConsumeReg(treeNode->AsOp()->gtOp1); // TODO: This does more than required. What is the proper way to say that this just consumes the value?
```

Comment:
I don't think there's anything like this the JIT today and I would guess that the only thing that's worth avoiding from `genCosumeReg` is unspilling. The safest way to avoid that is probably to put `tree->gtGetOp1()->SetRegOptional()` in `GT_SIDEEFFECT`'s lowering.

Reply:
I was not able to make this work. The last commit has what I have tried and the assert I got. Could you please take a look and suggest the proper way to do it?

Also, I assume I would need to call something else here instead `genConsumeReg`. What it would be?

Reply:
Sorry, I forgot to mention that. If it's reg optional then codegen has to check that the operand is really in a register - `if (treeNode->AsOp()->gtOp1->isUsedFromReg()) genConsumeReg(…`

Reply:
If I call `SetRegOptional` in https://github.com/dotnet/coreclr/pull/27554/files#diff-f36b3b4ad6feb9469b14f5ba2047cf74R352, I am getting `Assertion failed '!operand->isContained()' in 'My:Main(ref):int'` a long before I even get here. There is something wrong with the `SetRegOptional`. Am I calling it in the right place? Are there some other invariants that needs to be followed for `SetRegOptional` during `LinearScan::BuildNode`?

I am starting to think that trying to avoid the dummy spill here may not be worth it for the same reason why it is better to be conservative with the side-effects in this case.

Reply:
Right, this needs to use `genConsumeRegs` (note the plural Regs) instead of `isUsedFromReg`+`genConsumeReg`. But let me double check first to be sure there are no other issues.

Reply:
You also need to change the LSRA code to call `BuildOperandUses` instead of `BuildUse`:
```C++
case GT_KEEPALIVE:
    assert(dstCount == 0);
    srcCount = BuildOperandUses(tree->gtGetOp1());
    break;
```

> I am starting to think that trying to avoid the dummy spill here may not be worth it for the same reason why it is better to be conservative with the side-effects in this case.

Perhaps, I haven't seen the diffs to know how often unspilling may happen. But unlike the side effects story this is pretty safe. It's just that I haven't use reg optional in a while and I don't remember every little detail. Not to mention that I can't ever remember the `genConsumeReg`/`genConsumeRegs` distinction :grin:

Reply:
>  I'm not certain exactly what ordering constraints you want to impose, but I'm wondering if GTF_ORDER_SIDEEFF would be appropriate

Well, KeepAlive was a call to begin with and the safest thing to do is surely to make the intrinsic look like a call as well unless there are obvious problems with doing that. Considering that it doesn't return a value and thus ends up in its own statement it seems very unlikely that the more conservative `GTF_CALL` would have a negative impact.

Reply:
I agree that we should just use `GTF_CALL` here. Out current uses of `GTF_ORDER_SIDEEFF` are inconsistent, I hope I'll have time to clean them up sometime.

Reply:
Thanks @erozenfeld